src/config/ability.js

```diff
@@ -0,0 +1,25 @@
+import { AbilityBuilder } from '@casl/ability'
+
+export default function defineAbilitiesFor (ctx) {
+  let user = (ctx.state.isAuthenticated) ? ctx.state.user : {roles: ['public']}
+  let roleExists = (roleArray, roleName) => {
+    return roleArray.indexOf(roleName) !== -1
+  }
+
+  return AbilityBuilder.define((can, cannot) => {
+    switch (true) {
```

Comment:
We should refactor this switch-statement to a simple dictionary. I will show you how:

```js
const abilitiesMapping = {
  user: {
    abilities: [{ entity: 'User', functions: ['read', 'update'] }]
  },
  Subscriber: {
    abilities: [{ entity: 'Video', functions: ['read'] }]
  },
  Admin: {
    abilities: [
      { entity: 'Video', functions: ['read', 'update'] },
      { entity: 'User', functions: ['read', 'write', 'update'] },
      { entity: 'Post', functions: ['update', 'delete'] }
    ]
  }
}

export default function defineAbilitiesFor(ctx) {
  return AbilityBuilder.define((can, cannot) => {
    let user = ctx.state.isAuthenticated
      ? ctx.state.user
      : { roles: ['public'] }
    const roleAbilities = user.roles.map(role => abilitiesMappping[role])
    roleAbilities.forEach(roleAbility => {
      can(roleAbility.functions, entity)
    })
  })
}
```

With this approach: a) all role abilities will be applied. Here it might be possible if a user has multiple roles that access to an entity is not given (see `user.roles` can have 'user' and 'Subsriber' role but just 'user' abilities would be applied because of the break in line 13).